src/software/ai/hl/stp/skill/dribble/dribble_skill_fsm.h

```diff
@@ -105,112 +91,121 @@ struct DribbleFSM
+     * excessive dribbling
      *
-     * @param event DribbleFSM::Update
+     * @param event the Update event
      *
-     * @return if the ball possession should be lost
+     * @return if the ball should be lost
      */
     bool shouldLoseBall(const Update &event);
 
+    /**
+     * Guard that checks whether the excessive_dribbling_mode control param
+     * is set to TbotsProto::ExcessiveDribblingMode::TERMINATE
+     *
+     * @param event the Update event
+     *
+     * @return whether the excessive_dribbling_mode control param is set
+     * to TbotsProto::ExcessiveDribblingMode::TERMINATE
+     */
+    bool shouldExcessivelyDribble(const Update &event);
+
     auto operator()()
     {
         using namespace boost::sml;
 
-        DEFINE_SML_STATE(GetPossession)
+        DEFINE_SML_STATE(GetBallControl)
         DEFINE_SML_STATE(Dribble)
         DEFINE_SML_STATE(LoseBall)
+
         DEFINE_SML_EVENT(Update)
-        DEFINE_SML_GUARD(havePossession)
-        DEFINE_SML_GUARD(lostPossession)
+
+        DEFINE_SML_GUARD(haveBallControl)
+        DEFINE_SML_GUARD(lostBallControl)
         DEFINE_SML_GUARD(dribblingDone)
         DEFINE_SML_GUARD(shouldLoseBall)
-        DEFINE_SML_ACTION(startDribble)
+        DEFINE_SML_GUARD(shouldExcessivelyDribble)
+
         DEFINE_SML_ACTION(loseBall)
-        DEFINE_SML_ACTION(getPossession)
+        DEFINE_SML_ACTION(getBallControl)
         DEFINE_SML_ACTION(dribble)
 
         return make_transition_table(
             // src_state + event [guard] / action = dest_state
-            *GetPossession_S + Update_E[havePossession_G] / startDribble_A = Dribble_S,
-            GetPossession_S + Update_E[!havePossession_G] / getPossession_A,
-            Dribble_S + Update_E[lostPossession_G] / getPossession_A = GetPossession_S,
-            Dribble_S + Update_E[shouldLoseBall_G] / loseBall_A      = LoseBall_S,
-            Dribble_S + Update_E[!dribblingDone_G] / dribble_A,
-            Dribble_S + Update_E[dribblingDone_G] / dribble_A = X,
-            LoseBall_S + Update_E[!lostPossession_G] / loseBall_A,
-            LoseBall_S + Update_E[lostPossession_G] / getPossession_A = GetPossession_S,
-            X + Update_E[lostPossession_G] / getPossession_A          = GetPossession_S,
-            X + Update_E[!dribblingDone_G] / dribble_A                = Dribble_S,
-            X + Update_E / dribble_A                                  = X);
+            *GetBallControl_S + Update_E[haveBallControl_G] / dribble_A = Dribble_S,
+            GetBallControl_S + Update_E / getBallControl_A,
+
+            Dribble_S + Update_E[lostBallControl_G] / getBallControl_A = GetBallControl_S,
+            Dribble_S + Update_E[shouldLoseBall_G && !shouldExcessivelyDribble_G] /
+                            loseBall_A = LoseBall_S,
+            Dribble_S +
+                Update_E[shouldLoseBall_G && shouldExcessivelyDribble_G] / dribble_A = X,
```

Comment:
does it make sense to have a `stop_A` when we want to stop dribbling?

I thought I commented this earlier but it seems to have disappeared

Reply:
probably not worth it, we will switch to a new skill straight away after entering terminate state + we want to keep the dribbler on